README.md
# dart_codable

Implementing the Swift Codable Protocol using mirrors.

## Example
First of all import the package alongside the dart:convert package.

    import 'dart:convert';
    import 'package:dart_codable/dart_codable.dart';


Have your Class extend the Codable Class. Note that you have to pass down the constructor from the superclass because dart does not automatically create non-default constructors.

    class Person extends Codable{
      Person.decode(Map<String, dynamic> data) : super.decode(data);

      String name;
      int age;
      double score;
    }

Then simply use it:

    String jsonString = "{ \"name\": \"Marco\", \"age\": 24, \"score\": 99.9 }";

    Person me = Person.decode(json.decode(jsonString));

    print("Name: ${me.name}");
    print("Age: ${me.age}");
    print("Score: ${me.score}");
    print("");

    print(json.encode(me.encode()));

Which outputs 

    Name: Marco
    Age: 24
    Score: 99.9

    {"name":"Marco","age":24,"score":99.9}


CHANGELOG.md
## [0.2.0] - 10.05.2020

* Support nesting classes.

## [0.1.0] - 10.05.2020

* Release version 0.1.0.

## [0.0.3] - 10.05.2020

* Reformat the code and extend example.

## [0.0.2] - 10.05.2020

* Add doc Strings and update the description.

## [0.0.1] - 09.05.2020

* Basic Implementation of the Swift Codable Protocol.


test/.test_coverage.dart
// Auto-generated by test_coverage. Do not edit by hand.
// Consider adding this file to your .gitignore.

import 'dart_codable_test.dart' as dart_codable_test;

void main() {
  dart_codable_test.main();
}


test/implementations.dart
import 'package:dart_codable/dart_codable.dart';


class EncodableImp extends Encodable{
  EncodableImp({this.name, this.age, this.score});

  String name;
  int age;
  double score;
}

class DecodableImp extends Decodable{
  DecodableImp.decode(Map<String, dynamic> data) : super.decode(data);

  String name;
  int age;
  double score;
}

class CodableImp extends Codable{
  CodableImp.decode(Map<String, dynamic> data) : super.decode(data);

  String name;
  int age;
  double score;
}

test/dart_codable_test.dart
import 'dart:convert';
import 'package:dart_codable/dart_codable.dart';
import 'package:test/test.dart';
import 'implementations.dart';

void main() {
  String data = "{ \"name\": \"Marco\", \"age\": 24, \"score\": 99.9 }";
  String corruptedData =
      "{ \"notname\": \"Marco\", \"age\": 24, \"score\": 99.9 }";

  group("FieldDoesNotExistException", () {
    test('FieldDoesNotExistException contains field name', () {
      expect(FieldDoesNotExistException("testField").toString(),
          contains("testField"));
    });
  });

  group("Encodable tests", () {
    test('Encodable creates valid Json', () {
      EncodableImp encodable =
          EncodableImp(name: "Marco", age: 24, score: 99.9);
      Map<String, dynamic> encodedJson = encodable.encode();

      expect(encodedJson["name"], "Marco");
      expect(encodedJson["age"], 24);
      expect(encodedJson["score"], 99.9);
    });
  });

  group("Decodable tests", () {
    test('Deodable sets all fiels correctly', () {
      DecodableImp decodable = DecodableImp.decode(json.decode(data));

      expect(decodable.name, "Marco");
      expect(decodable.age, 24);
      expect(decodable.score, 99.9);
    });

    test('Deodable throws FieldDoesNotExistException', () {
      expect(() => DecodableImp.decode(json.decode(corruptedData)),
          throwsA(const TypeMatcher<FieldDoesNotExistException>()));
    });
  });

  group("Codable tests", () {
    test('Codable encode decode returns valid Json', () {
      CodableImp codable = CodableImp.decode(json.decode(data));
      Map<String, dynamic> encodedJson = codable.encode();

      expect(encodedJson["name"], "Marco");
      expect(encodedJson["age"], 24);
      expect(encodedJson["score"], 99.9);
    });
  });
}


lib/dart_codable.dart
library dart_codable;

import 'dart:mirrors';

/// Class implementing both the Decodable and the Encodable Protocol.
abstract class Codable extends Decodable with Encodable {
  /// Making the named constructor from Decodable available to children.
  Codable.decode(Map<String, dynamic> data) : super.decode(data);
}

/// Class implementing the Decodable Protocol.
abstract class Decodable {
  /// Initializes the child class with the data specified in the data Map.
  ///
  /// Iterate over the data map and initialize the fields of a class extending
  /// the Decodable Interface with the value provided.
  ///
  /// @param data Map of field name to initial value.
  /// @throws FieldDoesNotExistException If the field specified by a map
  /// key does not exist.
  Decodable.decode(Map<String, dynamic> data) {
    InstanceMirror mirror = reflect(this);
    data.forEach((key, value) {
      try {
        VariableMirror field = mirror.type.declarations[Symbol(key)];
        if (field.type.isSubtypeOf(mirror.type.superclass)) {
          ClassMirror fieldType =
              mirror.type.instanceMembers[Symbol(key)].returnType;
          mirror.setField(Symbol(key),
              fieldType.newInstance(Symbol("decode"), [value]).reflectee);
        } else {
          mirror.setField(Symbol(key), value);
        }
      } catch(_) {
        throw FieldDoesNotExistException(key);
      }
    });
  }
}

/// A custom Exception to be thrown if you try to initialize a field
/// that does not exist.
class FieldDoesNotExistException implements Exception {
  /// Set the non-existent field.
  FieldDoesNotExistException(this.field);

  /// Field name of the non-existent field.
  final String field;

  /// String representation of the exception.
  @override
  String toString() {
    return "FieldDoesNotExistException: Tried setting $field on class, "
        "which does not exist";
  }
}

/// Class implementing the Encodable Protocol.
abstract class Encodable {
  /// Encode the class into a Map of field name to value.
  ///
  /// Iterate over the all fields in the class and create a map with the
  /// field name as a key and the fields value as a value.
  ///
  /// @returns Map of field name to initial value.
  Map<String, dynamic> encode() {
    InstanceMirror mirror = reflect(this);
    Map<String, dynamic> encoded = Map<String, dynamic>();

    for (var v in mirror.type.declarations.values) {
      var name = MirrorSystem.getName(v.simpleName);
      if (v is VariableMirror && !v.isPrivate) {
        if (v.type.isSubtypeOf(mirror.type.superclass)) {
          encoded[name] = mirror
              .getField(Symbol(name))
              .invoke(Symbol("encode"), []).reflectee;
        } else {
          encoded[name] = mirror.getField(Symbol(name)).reflectee;
        }
      }
    }
    return encoded;
  }
}


example/main.dart
import 'dart:convert';

import 'package:dart_codable/dart_codable.dart';

class Person extends Codable {
  Person.decode(Map<String, dynamic> data) : super.decode(data);

  String name;
  int age;
  Grades grades;
}

class Grades extends Codable {
  Grades.decode(Map<String, dynamic> data) : super.decode(data);

  String subject;
  double grade;
}

String jsonString =
    "{ \"name\": \"Marco\", \"age\": 24, \"grades\": { \"subject\": \"Math\", \"grade\": 1.0 } }";
String corruptedJsonString =
    "{ \"notname\": \"Marco\", \"age\": 24, \"score\": 99.9 }";

void main() {
  try{
    Person me = Person.decode(json.decode(jsonString));

    print("Name: ${me.name}");
    print("Age: ${me.age}");
    print("Grades subject: ${me.grades.subject}");
    print("Grades grade: ${me.grades.grade}");
    print("");

    print(json.encode(me.encode()));
    print("");
  } catch(e) {
    print(e);
  }

  try {
    Person.decode(json.decode(corruptedJsonString));
  } catch (e) {
    print(e);
  }
}
